src/xrpld/app/misc/detail/LendingHelpers.cpp

```diff
@@ -0,0 +1,2052 @@
+ * - calculateRawLoanState: Computes theoretical values at full precision
+ * - constructRoundedLoanState: Builds state from actual rounded ledger values
+ *
+ * The interestDue field is derived from the other three values rather than
+ * stored directly, since it can be calculated as:
+ *   interestDue = totalValueOutstanding - principalOutstanding -
+ * managementFeeOutstanding
+ *
+ * This ensures consistency across the codebase and prevents copy-paste errors
+ * when creating LoanState objects from Loan ledger data.
+ */
+LoanState
+constructLoanState(
+    Number const& totalValueOutstanding,
+    Number const& principalOutstanding,
+    Number const& managementFeeOutstanding)
+{
+    // This implementation is pretty trivial, but ensures the calculations
+    // are consistent everywhere, and reduces copy/paste errors.
+    return LoanState{
+        .valueOutstanding = totalValueOutstanding,
+        .principalOutstanding = principalOutstanding,
+        .interestDue = totalValueOutstanding - principalOutstanding -
+            managementFeeOutstanding,
+        .managementFeeDue = managementFeeOutstanding};
+}
+
+LoanState
+constructRoundedLoanState(SLE::const_ref loan)
+{
+    return constructLoanState(
+        loan->at(sfTotalValueOutstanding),
+        loan->at(sfPrincipalOutstanding),
+        loan->at(sfManagementFeeOutstanding));
+}
+
+/*
+ * This function calculates the fee owed to the broker based on the asset,
+ * value, and management fee rate.
+ *
+ * Equation (32) from XLS-66 spec, Section A-2 Equation Glossary
+ */
+Number
+computeManagementFee(
+    Asset const& asset,
+    Number const& value,
+    TenthBips32 managementFeeRate,
+    std::int32_t scale)
+{
+    return roundToAsset(
+        asset,
+        tenthBipsOfValue(value, managementFeeRate),
+        scale,
+        Number::downward);
+}
+
+/*
+ * Given the loan parameters, compute the derived properties of the loan.
+ */
+LoanProperties
+computeLoanProperties(
+    Asset const& asset,
+    Number principalOutstanding,
+    TenthBips32 interestRate,
+    std::uint32_t paymentInterval,
+    std::uint32_t paymentsRemaining,
+    TenthBips32 managementFeeRate,
+    std::int32_t minimumScale)
+{
+    auto const periodicRate = loanPeriodicRate(interestRate, paymentInterval);
+    XRPL_ASSERT(
+        interestRate == 0 || periodicRate > 0,
+        "ripple::computeLoanProperties : valid rate");
+
+    auto const periodicPayment = detail::loanPeriodicPayment(
+        principalOutstanding, periodicRate, paymentsRemaining);
+
+    auto const [totalValueOutstanding, loanScale] = [&]() {
+        NumberRoundModeGuard mg(Number::to_nearest);
+        // Use STAmount's internal rounding instead of roundToAsset, because
+        // we're going to use this result to determine the scale for all the
+        // other rounding.
+
+        // Equation (30) from XLS-66 spec, Section A-2 Equation Glossary
+        STAmount amount{asset, periodicPayment * paymentsRemaining};
```

Comment:
what kind of difference would it make if `amount` is `Number` type? What does using `STAmount` give us - wouldn't `STAmount` still always use `Number` to do the normalization since `getSTNumberSwitchover()` is always true?